HIP/hip-1056.md

```diff
@@ -0,0 +1,1952 @@
+  * transfers performed by the transaction, or by a smart contract it calls, or
+  * by the creation of threshold records that it triggers.
+  */
+  proto.TransferList transfer_list = 6;
+
+  /**
+  * All Token transfers as a result of this transaction.
+  */
+  repeated proto.TokenTransferList token_transfer_lists = 7;
+
+  /**
+  * All token associations implicitly created while handling this transaction.
+  */
+  repeated proto.TokenAssociation automatic_token_associations = 8;
+
+  /**
+  * List of accounts with the corresponding staking rewards paid as a result of
+  * a transaction.
+  */
+  repeated proto.AccountAmount paid_staking_rewards = 9;
+
+  /**
+  * Congestion pricing multiplier at the time the transaction was executed.
+  */
+  uint64 congestion_pricing_multiplier = 10;
+}
+```
+
+<aside>
+🚨 **Open Decision**: Decide on string error message or a integer detailed error code?
+
+</aside>
+
+<aside>
+🚨 **Open Question**: We discussed if TransferList should be included or not here. I don’t remember exactly why it has
+to be. We should double check if it has to be and can not be removed and if it has to be here then document why.
+ - The Mirror Node MAPI exposes the transfer list, and it’s used by clients of mirror node.
+ - There are some transfers only exposed in the transfer list currently.
+</aside>
+
+#### Transaction Output
+
+Output data from a transaction data that is neither present in the transaction nor stored in state. Not all transactions
+have output, each one that does has a corresponding output message type with any data it needs to include.
+
+<aside>
+🚨 **Open Task: Update section with missing TransactionOutput objects**.
+</aside>
+
+```protobuf
+message TransactionOutput {
+  oneof transaction {
+    /**
+    * Output from a crypto transfer transaction.
+    */
+    CryptoTransferOutput crypto_transfer = 1;
+
+    /**
+    * Output from a utilPrng transaction to request a
+    * deterministic pseudo-random number.
+    */
+    UtilPrngOutput util_prng = 2;
+
+    /**
+    * Output from a contract call transaction.
+    */
+    CallContractOutput contract_call = 3;
+
+    /**
+    * Output from an ethereum call transaction.
+    */
+    EthereumOutput ethereum_call = 4;
+
+    /**
+    * Output from a contract create transaction.
+    */
+    CreateContractOutput contract_create = 5;
+
+    /**
+    * Output from a schedule create transaction that executed
+    * immediately on creation.
+    */
+    CreateScheduleOutput create_schedule = 6;
+
+    /**
+    * Output from a schedule sign transaction that resulted in
+    * executing the scheduled transaction.
+    */
+    SignScheduleOutput sign_schedule = 7;
+  }
+}
+```
+
+```protobuf
+message CryptoTransferOutput {
+  /**
+  * Custom fees assessed during a CryptoTransfer.
+  * <p>
+  * These fees SHALL be present in the full transfer list for the transaction.
+  */
+  repeated proto.AssessedCustomFee assessed_custom_fees = 1;
```

Comment:
HIP-991 will enable custom fees for topics, and most likely this field will be used to externalize the custom fees charged for a consensus submit message transaction. Should we take the opportunity and rename `CryptoTransferOutput` to be generic?

Reply:
It is probably more appropriate, as part of `HIP-991`, to add `assessed_custom_fees` to `SubmitMessageOutput` and include that message in `TransactionOutput` (to be generated when a consensus submit charges custom fees).

Reply:
I think Xin is saying that since now `AssessedCustomFee` will appear in at least two transaction types and probably more in the future, instead of adding it to each different output we should consider pulling it up to the `TransactionResult` since it's more common. This is important to consider now, not in HIP-991 since it's a breaking change later.

Reply:
The design thus far has been to add outputs for a given transaction to a transaction-specific output (for the few transactions that have outputs).

Changing that, at least in this case, would make it much harder to filter the block stream to, for example, only show the activity for consensus service transactions, as you would _also_ need to read all of a "generic" set of outputs and tie them to specific transactions (so the filter would need to be stateful).

It's certainly an option, but I would suggest that maximizing the options for stateless filters in the Block Stream is worth the minor duplication of concepts (no _data_ would be duplicated, of course).

Reply:
Once an output starts appearing on more than one transaction type and if we think it might appear in more we should consider moving it to `TransactionResult`. This is already the case for fields in the `TransactionResult` and we'd be following the same model. `TokenTransferList` appears in the result but it only applies to `TokenReject`, `TokenAirdrop`, and `CryptoTransfer`. `TokenAssociation` in the result only applies to `TokenCreate` and `CryptoTransfer`. There is strong precedent to consider `AssessedCustomFee` to be the same.

Reply:
That's not how I view it, but it's entirely possible my view is not correct.

As I view these items, `TransactionResult` is _mostly_ items that are _global_ to all transactions, as well as a few items that are currently expected by Mirror Node in the record stream but would be difficult to move into individual transaction results.
Ideally, in my opinion, we _would_ move those "extra" items into individual transaction results, or state changes, and `TransactionResult` would be reduced to just the properly "global" items like status, consensus timestamps, congestion multiplier, and transaction fee; perhaps paid staking rewards (not sure if that's universal or specific) as well.

Reply:
These are valid considerations.
 
Personally I wish `TransactionResult` was not needed or at the least was exactly as @jsync-swirlds noted and didn't include `transfer_list`, `token_transfer_lists` or `automatic_token_associations` which are present because they are results from transactions that differ from their usual associated transaction type.

@steven-sheehy and @xin-hedera are right that there's a precedence here to add  `assessed_custom_fees`.
My concern is what else does that open the door and how do we ensure consistency and ability to manage future pivots.
However, if we rely on future move of items in a specific output type to the general `TransactionResult` we might require multiple deprecations of those fields to avoid duplication. For instance if Block Stream were already in use we'd now be considering how to stop using `CryptoTransferOutput.assessed_custom_fees` and how to start using `TransactionResult.assessed_custom_fees`. This will certainly happen again for other fields as HAPI continues to mature.

Reply:
@jsync-swirlds how does the inclusion of `assessed_custom_fees` impact filtering functionality in a way different from say `token_transfer_lists` or `automatic_token_associations`?

Reply:
@steven-sheehy what's the downside if `assessed_custom_fees` appeared in both `CryptoTransferOutput` and `TokenAirdropOutput`? Is it just the duplication the protobuf and the parsing consideration to look in more than one object for the same value depending on transaction type? Is this better as it ensure transaction types are more self contained?

Reply:
Token transfer list and automatic associations are both present elsewhere; their inclusion here is duplicative, and actually _very hard_ to remove if a redaction were to become necessary.  Both of the mentioned current fields are already a problem for redaction and filtering.  Basically we can redact whole messages and preserve the block proof, but if we _change_ a single message (say, to remove custom fees related to a redacted message), the block proof will not validate, so we could remove the message that related to a custom fee (perhaps an HCS submit), but could not remove the related custom fees from this message, which limits how thoroughly we could comply with a redaction request.
For general filtering it's possible to filter out the state change, transaction, and outputs that duplicate the data here; but this data would likely remain in the stream unless this output message was also filtered (which would require a complex filter, and associated higher charge).
Part of the reason for being very careful to minimize fields in this message is to reduce cases where redaction is limited and filtering is more complex.